README.md
# What is this?
It is a module that expose data read from an MPU6050 sensor (acceleration + gyroscope) using a Raspberry Pi board.

PRs are welcome.

# Instructions

### Requirements:
 - Raspberry Pi board
 - MPU6050 module
 - Node.js >= 10 (LTS)

### Setting up the Raspberry Pi board:
This module requires a proper setup of a Raspberry Pi board. A detailed walkthrough on how to prepare the Raspberry Pi setup can be found [here](./PI-SETUP.md).

### To install locally: 
1. Clone the repository in a directory `[directory]`
2. `cd [directory]`
3. `npm install`
4. `npm link`
5. `read-mpu --help`
6. If you want to connect to AWS generate and install the certificates as described [here](./PI-SETUP.md#certificates) 
7. Enjoy.

### To uninstall:
1. Go to repo directory
2. `npm unlink`
3. Remove the repo directory

### To run tests:
The package includes a **full test coverage**.
Go inside the repo directory and run: 

`npm test`

Once executed you can find the coverage report in a neat html format:
1. `cd coverage`
2. open with a browser the file `index.html`

### Usage:
The module can:
- be installed as a component in another node program
- be used from the CLI by invoking `read-mpu` with the proper options. Type `read-mpu --help` for a complete list of available commands

### Notes:
The code is fully commented.
All the functions and helpers are documented using JSHint conventions. That will trigger autocompletion and code hints in code editors like Visual Studio Code.

CHANGELOG.md
## 1.0.0 (Mar 28, 2019)

Initial release
  - mpu6050 module
  - read-mpu CLI command

test/test.js
/* 
  <mpu6050 tests>
  Copyright (C) <2019>  <Luca Silvestri>

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

/* eslint-env node, mocha */
/* eslint quotes: ["error", "single"]*/

var assert = require('chai').assert;
var Sensor = require('../mpu6050');

const DEFAULT_TEST_LIMIT = 5;

describe('Tries a sync read of the sensor and check for result', function(){
  var sensorInstance = new Sensor(0x68);
  var data = sensorInstance.readSync();
  //Repeats the read for cold starts
  data = sensorInstance.readSync();
  console.log(data);
  it('Tests if data is properly formatted', function(){
    assert.property(data,'gyro', 'Does not have gyro property');
    assert.property(data,'accel', 'Does not have accel property');
    assert.property(data,'rotation', 'Does not have rotation property');
    assert.property(data,'temp', 'Does not have temp property');
    assert.nestedProperty(data,'gyro.x', 'Does not have gyro.x property');
    assert.nestedProperty(data,'gyro.y', 'Does not have gyro.y property');
    assert.nestedProperty(data,'gyro.z', 'Does not have gyro.z property');
    assert.nestedProperty(data,'accel.x', 'Does not have accel.x property');
    assert.nestedProperty(data,'accel.y', 'Does not have accel.y property');
    assert.nestedProperty(data,'accel.z', 'Does not have accel.z property');
    assert.nestedProperty(data,'rotation.x', 'Does not have rotation.x property');
    assert.nestedProperty(data,'rotation.y', 'Does not have rotation.y property');
  });

  it('Tests calibration', function(){
    //Calibrates sensor
    sensorInstance.calibrateSensor(data);
    //Read the calibrated data from the sensor
    let calibratedData = sensorInstance.readSync();
    console.log('Calibrated data:');
    console.log(calibratedData);
    assert.isTrue(calibratedData.accel.x < 0.1 && calibratedData.accel.x > -0.1, 'Accel x is out of calibration limits');
    assert.isTrue(calibratedData.accel.y < 0.1 && calibratedData.accel.y > -0.1, 'Accel y is out of calibration limits');
    assert.isTrue(calibratedData.accel.z < 0.1 && calibratedData.accel.z > -0.1, 'Accel z is out of calibration limits');
  });

  it('Tests read asyncronously', function(done){
    //Read asyncronously
    sensorInstance.readAsync(function(err, asyncData){
      if(err){
        done(err);
      } else {
        console.log('Async data: ');
        console.log(asyncData);
        assert.property(asyncData,'gyro', 'Does not have gyro property');
        assert.property(asyncData,'accel', 'Does not have accel property');
        assert.property(asyncData,'rotation', 'Does not have rotation property');
        assert.property(asyncData,'temp', 'Does not have temp property');
        assert.nestedProperty(asyncData,'gyro.x', 'Does not have gyro.x property');
        assert.nestedProperty(asyncData,'gyro.y', 'Does not have gyro.y property');
        assert.nestedProperty(asyncData,'gyro.z', 'Does not have gyro.z property');
        assert.nestedProperty(asyncData,'accel.x', 'Does not have accel.x property');
        assert.nestedProperty(asyncData,'accel.y', 'Does not have accel.y property');
        assert.nestedProperty(asyncData,'accel.z', 'Does not have accel.z property');
        assert.nestedProperty(asyncData,'rotation.x', 'Does not have rotation.x property');
        assert.nestedProperty(asyncData,'rotation.y', 'Does not have rotation.y property');
        done();
      }
    });
  });
  describe('Tests poolSensorData', function(){
    this.timeout(120000);
    it('poolSensorData: all the options', function(done){
      let count = 0;
      sensorInstance.poolSensorData({interval: 500, limit: 10}, function(err, data){
        if(err){
          done(err);
        } else {
          console.log(data[0]);
          assert.strictEqual(data[1].interval, 500);
          assert.strictEqual(data[1].limit, 10);
          assert.strictEqual(typeof data[2], 'function');
          count = count + 1;
          if(count >= 10){
            done();
          }
        }
      }, true);  
    });

    it('poolSensorData: number', function(done){
      let count = 0;
      sensorInstance.poolSensorData(3000, function(err, data){
        if(err){
          done(err);
        } else {
          assert.strictEqual(data[1].interval, 3000);
          assert.strictEqual(typeof data[2], 'function');
          count = count + 1;
          if(count >= DEFAULT_TEST_LIMIT){
            done();
          }
        }
      }, true);  
    });

    it('poolSensorData: options is null', function(done){
      let count = 0;
      sensorInstance.poolSensorData(null, function(err, data){
        if(err){
          done(err);
        } else {
          assert.strictEqual(data[1].interval, 1000);
          assert.strictEqual(typeof data[2], 'function');
          count = count + 1;
          if(count >= DEFAULT_TEST_LIMIT){
            done();
          }
        }
      }, true);
    });

    it('poolSensorData: function as first argument', function(done){
      let count = 0;
      sensorInstance.poolSensorData(function(err, data){
        if(err){
          done(err);
        } else {
          assert.strictEqual(data[1].interval, 1000);
          assert.strictEqual(typeof data[2], 'function');
          count = count + 1;
          if(count >= DEFAULT_TEST_LIMIT){
            done();
          }
        }
      }, null, true);
    });

    it('poolSensorData: string as first argument', function(done){
      sensorInstance.poolSensorData('hello', function(err){
        if(err){
          assert.strictEqual(err.name, 'WRONG_PARAM_TYPE', 'Wrong error type');
          done();
        } else {
          assert.ok(false, 'Did not got an error when supplied string as option value');
          done();
        }
      }, true);
    });

    it('poolSensorData: only callback as argument', function(done){
      let count = 0;
      sensorInstance.poolSensorData(function(err, data){
        if(err){
          done(err);
        } else {
          assert.strictEqual(data[1].interval, 1000);
          assert.strictEqual(typeof data[2], 'function');
          count = count + 1;
          if(count >= DEFAULT_TEST_LIMIT){
            done();
          }
        }
      }, null, true);
    });  
    
    it('poolSensorData: no callback provided', function(done){
      assert.doesNotThrow(() => sensorInstance.poolSensorData({interval: 3000, limit: 2}, null, true));      
      done();
    }); 
    
    it('poolSensorData: no proper callback provided', function(done){
      assert.doesNotThrow(() => sensorInstance.poolSensorData({interval: 3000, limit: 2}, 'no proper callback', true));      
      done();
    });     
    
  });
});

mpu6050.js
#!/usr/bin/env node

/* 
  <mpu6050 module>
  Copyright (C) <2019>  <Luca Silvestri>

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

var i2c = require('i2c-bus');
var i2cMPU6050 = require('i2c-mpu6050');

const DEFAULT_POOL_INTERVAL = 1000;
const DEFAULT_TEST_LIMIT = 5;

/**
 *Describes a MPU6050 sensor
 *
 * @class MPU6050
 */
class MPU6050 {
  /**
   *Creates an instance of MPU6050.
   * @param {Number} address - Sensor address number in HEX notation
   * @memberof MPU6050
   */
  constructor(address){
    var i2c1 = i2c.openSync(1);
    this.sensor = new i2cMPU6050(i2c1,address);
    //Executes an initial read to initialize the sensor after a cold boot.
    this.readSync();
  }

  /**
   * Read data from sensor syncrounously
   *
   * @returns An object containing 
   * @memberof MPU6050
   */
  readSync(){
    var data = this.sensor.readSync();
    return data;
  }

  /**
   * Read data from sensor asyncrounously
   *
   * @param {function(Error, object)} callback The callback(err, data) function 
   * @memberof MPU6050
   */
  readAsync(callback) {
    this.sensor.read(callback);
  }

  /**
   *Perform a sensor calibration with data from a static measure
   *
   * @param {{object}} staticMeasure A sensor read when it's static (not moving)
   * @memberof MPU6050
   */
  calibrateSensor(staticMeasure){
    //Defines a calibration object
    let calibration = {
      accel: {},
      gyro: {}
    };
    //Evaluates the calibration values for acceleration
    for (const key in staticMeasure.accel) {
      /* istanbul ignore else  */
      if (staticMeasure.accel.hasOwnProperty(key)) {
        const element = staticMeasure.accel[key];
        calibration.accel[key] = -element;
      }
    }
    this.sensor.calibrateAccel(calibration.accel);
  }

  /**
   * Pools sensor data at a given interval
   *
   * @param {{interval:Number, limit:Number}} opts Options. Interval default is 1000s
   * @param {function(Error,Any)} callback Callback function (err, data)
   * @param {boolean} test True if it is used in a test suite
   * @returns
   * @memberof MPU6050
   */
  poolSensorData(opts, callback, test){
    //Check parameters
    if(!opts){
      //opts is null. Sets default interval
      opts = {
        interval: DEFAULT_POOL_INTERVAL
      };      
    } else {
      switch (typeof opts) {
        case 'function':
          //callback passed as first argument. Set interval as default
          callback = opts;
          opts = {
            interval: DEFAULT_POOL_INTERVAL
          };          
          break;
        case 'number':
          //puts the supplied interval in the opts object
          opts = {
            interval: opts
          };
          break;
        case 'string':
          //Rise an error and exits
          let err = new Error('poolSensorData: opts parameter has to be a number or an object');
          err.name = 'WRONG_PARAM_TYPE';
          return callback(err);
        default:
          break;
      }
    }
    if(!callback){
      //If callback is not there, then substitute with a dummy function
      callback = () => {return;};
    } else {
      switch (typeof callback) {
        case 'function':
          //callback is a function. Go on.          
          break;      
        default:
          //callback is anything but a function. Substitute with dummy function
          callback = () => {return;}; 
          break;
      }
    }
    //Check if the method is called in a test scenario (e.g. Mocha)
    if(test){
      //Check if there is a previously specified limit in opts object
      if(!opts.limit){
        opts.limit = DEFAULT_TEST_LIMIT;
      }
    }
    let count = 0;
    //Set an interval timer to read the sensor at every given interval
    let intervalTimer = setInterval(() => {
      this.sensor.read(function(err, data){
        //Deal with limited numbers of consecutive reads.
        /* istanbul ignore else  */
        if(opts.limit){
          count = count + 1;
          //Checks if the limit number of consecutive reads have been reached
          if(count >= opts.limit){
            //Cancels the interval timer
            clearInterval(intervalTimer);
          }
        }
        //Return the read result
        if(err){
          return callback(err, null);
        } else {
          return callback(null, [data, opts, callback]);
        }        
      });
    },opts.interval);
  }
}

module.exports = MPU6050;

cli.js
#!/usr/bin/env node

/* 
  <read-mpu>
  Copyright (C) <2019>  <Luca Silvestri>

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

var program = require('commander');
const chalk = require('chalk');
var awsIot = require('aws-iot-device-sdk');
var pkg = require('./package.json');
var Sensor = require('./mpu6050');
var Table = require('cli-table3');
const os = require('os');

const config = require('./config/config.json');

var colorsOn = true;  //Default output is coloured => true
var trasmitInterval = 1000; //Default value for Interval between data transmit (ms)
var sendData = false; //True when is time to send aggregated data to AWS IoT
var iot = {
  'host': config.cliOptions.host,
  'keyPath': os.homedir() + config.cliOptions.privateKeyPathRelativeToHome,
  'certPath': os.homedir() + config.cliOptions.certificatePathRelativeToHome,
  'caPath': os.homedir() + config.cliOptions.caPathRelativeToHome,
  'clientId': 'read-mpu',
};
const TOPIC_PREFIX = 'device/mpu6050';

//Extracts cli program name from package.json
var cmd = pkg.name;
if (pkg.bin) {
    for (var key in pkg.bin) {
        cmd = key;
        break;
    }
}
//TODO: Add a flag for custom sensor-ID
//Initialize program object
program
  .name(cmd)
  .version('1.0.0')
  .option('-c, --connect', 'Connects to AWS IoT Core')
  .option('-i, --interval [value]', 'Interval between sensor read (ms)', parseInt)
  .option('-l, --limit [value]', 'Limit the number of reads', parseInt)
  .option('-t, --trasmit-interval [value]', 'Interval between data transmit (s)', parseInt)
  .option('-a, --all', 'Prints all the reads from the sensor')
  .option('-H, --host [value]', 'AWS IoT Endpoint')
  .option('--private-key [path]','Private key path')
  .option('--public-key [path]','Public key path')
  .option('--certificate [path]','Device certificate in PEM format')
  .option('--ca [path]','CA certificate')
  .option('--no-colors','Turns off colored output')
  .parse(process.argv);

/*
program.on('--help', function(){
  console.log('read-mpu  Copyright (C) 2019  Luca Silvestri');
  console.log('This program comes with ABSOLUTELY NO WARRANTY.');
  console.log('This is free software, and you are welcome to redistribute it');
  console.log('under GPLv3 License or above');
});
*/

/**
 *Initialize the program
 *
 */
function init(){
  //Checks for no-color option
  if(program.noColors){
    colorsOn = false;
  }
  //Checks for interval option
  if(program.interval){
    //Defaults trasmitInterval to interval provided value
    trasmitInterval = program.interval;
  }
  //Checks for tramsit interval option
  if(program.trasmitInterval){
    trasmitInterval = program.trasmitInterval * 1000;
  }
  //TODO: Change default values for certs/keys when provided as arguments
  //Prints start message
  startMessage();
  //Check if connection to AWS IoT is needed
  if(program.connect){
    console.log(styleIt('magenta','Connecting to AWS IoT...'));
    connect2IotCore((err, device)=>{
      if(err){
        //Something bad happened at connection. Just state the error and quit.
        console.error(styleIt('red','Error connecting AWS IoT:'));
        console.error(styleIt('red',err.message));
        throw err;
      } else {
        //Everything went fine. Launch main()
        console.log(styleIt('green','CONNECTED!'));
        return main(device);
      }
    });
  } else {
    //No connection required. Launch main()
    return main();
  }
}

/**
 *Prints out the start message
 *
 * @returns
 */
function startMessage(){
  console.log('');
  console.log('*****************************************************************');
  console.log(styleIt('green', 'read-mpu  Copyright (C) 2019  Luca Silvestri'));
  console.log('This program comes with ABSOLUTELY NO WARRANTY.');
  console.log('This is free software, and you are welcome to redistribute it');
  console.log('under GPLv3 License or above');
  console.log('*****************************************************************');
  return;
}
/**
 *Connect the sensor to AWS IoT
 *
 * @param {function(Error,Object)} callback Callback(err,data)
 * @returns
 */
function connect2IotCore(callback){
  //Check if host has been specified
  if(!program.host && !config.cliOptions.host){
    let hostError = new Error('Failed to provide a valid hostname');
    hostError.name = 'NO_HOST';
    return callback(hostError, null);
  } else {
    //Check wether host has been specified at cli input
    if(program.host){
      //Priority at command line options      
      iot.host = program.host;
    }
    let device = awsIot.device(iot);
    device.on('connect', () => {
      console.log('AWS IoT connected!');
      return callback(null, device);
    });
    device.on('error', (err) => {
      return callback(err, null);
    });
  }
}

/**
 *Formats the output string with a given color
 *
 * @param {string} color A chalk-related color code
 * @param {string} string The string that has to be formatted
 * @returns
 */
function styleIt(color, string){
  if(colorsOn){
    return chalk[color](string);
  }
  else {
    return string;
  }
}

/**
 *Read sensor data
 *
 * @param {number} interval Interval between sensor read (ms)
 * @param {number} limit Limit the number of reads
 * @param {awsIot.device} device An instance of mqtt client connected to AWS IoT
 */
function readData(interval, limit, device){
  var sensor = new Sensor(0x68);
  var table = new Table({
    colWidths: [20,20,20,20]
  });
  var maxData = {ax:0, ay:0, az:0, a:0};
  //Calibrates the sensor with data at-rest (to detract the effect of G accelleration)
  var intialData = sensor.readSync();
  sensor.calibrateSensor(intialData);
  // eslint-disable-next-line quotes
  //console.log(styleIt('yellow', `ax                     ay                      az`));
  table.push([styleIt('yellow','ax'),styleIt('yellow','ay'),styleIt('yellow','az'), styleIt('yellow','|a|')]);
  console.log(table.toString());
  // eslint-disable-next-line quotes
  //console.log(styleIt('yellow', `-----------------------------------------------------------`));
  sensor.poolSensorData({interval: interval, limit: limit}, function(err, data){
    if(err){
      console.error(styleIt('red', 'An error occurred while reading sensor data.'));
      console.error(styleIt('red', err.message));
      return;
    } else {
      //console.log(styleIt('yellow', `${data[0].accel.x}         ${data[0].accel.y}         ${data[0].accel.z}`));
      let ax = data[0].accel.x,
          ay = data[0].accel.y,
          az = data[0].accel.z,
          a  = Math.sqrt(ax**2 + ay**2 + az**2);
      table[0] = [styleIt('yellow', ax),styleIt('yellow', ay),styleIt('yellow', az), styleIt('yellow', a)];
      //Checks if maxData has to be updated
      if(maxData.a <= a){
        maxData = {ax:ax, ay:ay, az:az, a:a};
      }
      //Check if data has to be sent
      if(sendData && device){
        //Sends data
        table[1] = ['SENDING DATA'];
        //TODO: implement --all data flag
        let dataToSend = maxData;
        //TODO: Personalize the topic
        device.publish(TOPIC_PREFIX, JSON.stringify(dataToSend));
        //Resets send data flag and the maxData object
        sendData = false;
        maxData = {ax:0, ay:0, az:0, a:0};
        console.log(table.toString());
        table.splice(1,1);
      } else {
        console.log(table.toString());
      }      
      return;
    }
  });
}

/**
 *Sets a timer to evaluate when is time to send data to AWS IoT
 *
 * @param {number} sendDataInterval Interval in ms
 */
function startSendDataTimer(sendDataInterval){
  setInterval(() => {
    sendData = true;
  }, sendDataInterval);
}

/**
 * Main procedure
 *
 * @param {awsIot.device} device An instance of mqtt client connected to AWS IoT
 */
function main(device){
  startSendDataTimer(trasmitInterval);
  readData(program.interval, program.limit, device);
}

init();